### PR TITLE
add drainable mechanism

### PR DESCRIPTION
**Pull request by nglglhtr**

### CODE REVIEW

File: contracts/common/misc/Drainable.sol [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+import {IERC20} from "openzeppelin-solidity/contracts/token/ERC20/IERC20.sol";
+import {IERC721} from "openzeppelin-solidity/contracts/token/ERC721/IERC721.sol";
+
+
+contract Drainable is Governable {
+
+  constructor (address _governance) public Governable(_governance) {}
+
+  function drainErc20 (
+    address[] calldata tokens,
```
Comment (atvanguard) [inline contracts/common/misc/Drainable.sol]: Let's also accept the `uint256[] amount`. Better to be flexible. Same for the next 2 functions.

File: contracts/common/misc/Drainable.sol [hunk 2]
```diff
@@ -0,0 +1,38 @@
+pragma solidity ^0.5.2;
+
+import { Governable } from "../governance/Governable.sol";
+
+import {IERC20} from "openzeppelin-solidity/contracts/token/ERC20/IERC20.sol";
+import {IERC721} from "openzeppelin-solidity/contracts/token/ERC721/IERC721.sol";
+
+
+contract Drainable is Governable {
```
Comment (atvanguard) [inline contracts/common/misc/Drainable.sol]: Inherit from `DepositManagerStorage` instead. Proxy upgradeable contracts need to have same memory layout.

File: contracts/common/misc/Drainable.sol [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+contract Drainable is DepositManagerStorage {
+
+  function drainTokens(
+    address[] calldata tokens,
+    uint256[] calldata values,
+    address destination
+  ) external onlyGovernance {
+    for (uint256 i = 0; i < tokens.length; i++) {
+      if (registry.isERC721(tokens[i])) {
```
Comment (atvanguard) [inline contracts/common/misc/Drainable.sol]: Let's remove this. Someone could send unmapped tokens to this contract as well. Have separate functions for draining 20/721s.

**Comment by nglglhtr** [comment]
@atvanguard Would you suggest tracking the deposited ether in DepositManagerStorage contract? https://github.com/maticnetwork/contracts/blob/79b5a190be09929def4ac5a6ffa2aee27efa954b/contracts/root/depositManager/DepositManager.sol#L152, since the deposited value is directly transferred to MaticWETH contract

File: contracts/root/depositManager/DepositManager.sol [hunk 4]
```diff
@@ -16,12 +16,13 @@ import {IDepositManager} from "./IDepositManager.sol";
...
 import {Lockable} from "../../common/mixin/Lockable.sol";
-
+import {Drainable} from "../../common/misc/Drainable.sol";
 contract DepositManager is
     DepositManagerStorage,
     IDepositManager,
     IERC721Receiver,
-    ContractReceiver
+    ContractReceiver,
+    Drainable
```
Comment (atvanguard) [inline contracts/root/depositManager/DepositManager.sol]: The entire point of this exercise was to replace the implementation of depositManager and then drain. Please don't inherit from `Drainable`. - In the test deploy a new instance of drainable, update address in depositManagerProxy and then call drain functions :)

File: test/helpers/deployer.js [hunk 5]
```diff
@@ -169,6 +169,17 @@ class Deployer {
     return this.depositManager
   }
 
+  async deployDrainable() {
+    this.drainable = await contracts.Drainable.new()
+    await this.depositManagerProxy.updateImplementation(this.drainable.address)
+    await this.updateContractMap(
```
Comment (atvanguard) [inline test/helpers/deployer.js]: Why is this required?

File: test/root/Drainable.test.js [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+      const values = []
+      tokens.push(testToken.address)
+      values.push(amount.toString())
+      drainable = await deployer.deployDrainable()
+      const data = drainable.contract.methods.drainErc20(tokens, values, destination).encodeABI()
+      await this.contracts.governance.update(
+        drainable.address,
+        data
+      )
+      utils.assertBigNumberEquality(await testToken.balanceOf(destination), amount)
```
Comment (atvanguard) [inline test/root/Drainable.test.js]: Did you remove the assertion where balanceOf deposit Manager should now be 0. Also can we test for partial amount? Like if DepositManager has 10 we should drain only 7.

Comment (nglglhtr) [reply test/root/Drainable.test.js]: yes, was that required? assertion for the user to get the queried amount of tokens should be enough, no?

File: test/helpers/deployer.js (hunk 5 above)
Comment (nglglhtr) [reply test/helpers/deployer.js]: Just for consistency, isn't a requirement for this. Would you suggest removing it?

File: test/root/Drainable.test.js (hunk 6 above)
Comment (nglglhtr) [reply test/root/Drainable.test.js]: yes, will add for partial amount too

File: test/helpers/deployer.js (hunk 5 above)
Comment (atvanguard) [reply test/helpers/deployer.js]: There is no question of consistency in contracts. Gotta save all storage and gas possible.

File: test/root/Drainable.test.js (hunk 6 above)
Comment (atvanguard) [reply test/root/Drainable.test.js]: > assertion for the user to get the queried amount of tokens should be enough, no
Isn't that just half the assertion? Better to write a complete story.
